File: reference/protocol/src/main/java/com/nagada/pulse/protocol/SyncResponse.java
package com.nagada.pulse.protocol;

import java.util.List;
import java.util.Map;

/**
 * Sync response from server: acknowledgments + new server events.
 */
public class SyncResponse {
    public List<String> successClientEventIds;
    public List<ServerEvent> newServerEvents;
    public int nextHeartbeatMs;
    public Map<String,String> errorClientEventIds;
    public SyncResponse() {
    }

    public SyncResponse(
        List<String> successClientEventIds, 
        List<ServerEvent> newServerEvents,
        int nextHeartbeatMs,
        Map<String,String> errorClientEventIds) {
        this.successClientEventIds = successClientEventIds;
        this.newServerEvents = newServerEvents;
        this.nextHeartbeatMs=nextHeartbeatMs;
        this.errorClientEventIds=errorClientEventIds;
    }
    //add getters 
    public List<String> getSuccessClientEventIds() {
        return successClientEventIds;
    }
    public List<ServerEvent> getNewServerEvents() {
        return newServerEvents;
    }
    public int getNextHeartbeatMs() {
        return nextHeartbeatMs;
    }
    public Map<String, String> getErrorClientEventIds() {
        return errorClientEventIds;
    }
}

File: client/dart/nagada_client/lib/core/models/sync_response.dart
import 'package:json_annotation/json_annotation.dart';
import 'server_event.dart';

part 'sync_response.g.dart';

/// Data Transfer Object for a successful sync response from the server.
@JsonSerializable(explicitToJson: true)
class SyncResponse {
  /// List of client event IDs that the server has successfully processed.
  final List<String> successClientEventIds;

  /// List of new events from the server that the client needs to apply.
  final List<ServerEvent> newServerEvents;

  final Map<String,String> errorClientEventIds;

  /// nextHeartbeatMs
  final int nextHeartbeatMs;

  SyncResponse({
    required this.successClientEventIds,
    required this.newServerEvents,
    required this.nextHeartbeatMs,
    required this.errorClientEventIds,
  });

  /// Creates a new `SyncResponse` from a JSON map.
  factory SyncResponse.fromJson(Map<String, dynamic> json) =>
      _$SyncResponseFromJson(json);

  /// Converts this `SyncResponse` to a JSON map.
  Map<String, dynamic> toJson() => _$SyncResponseToJson(this);
}